### DISCUSSION TITLE
Basic Question: How to configure backend URL

### DISCUSSION THREAD

**Discussion by MasterBehind** [gabrieldwight/Whatsapp-Business-Cloud-Api-Net#56] [Q&A] [answered]
Hi,
I have setup all the required in WhatsApp Meta Developer/Business center and the last thing is to configure the Webhook, for that it is asking me to put the callback URL (End Point). So i used the sample project here and published the WhatsAppBusinessCloudAPI.Web to the following URL http://www.xyz.com/endpoint/
in this case, what should i put the callback URL?? the full URL

**Comment by gabrieldwight** [comment]
Hi, if you are testing the webhook by debugging the project you will need to use dev tunnels from visual studio or you can use conveyor extension to test the backend URL since it needs to receive verification from meta. 

The callback URL is the full URL pointing to the webhook route but it will need to go through the same verification process from meta.

**Comment by MasterBehind** [comment]
thanks a lot, but what is the "**ull URL pointing to the webhook route**" in my case assuming that i published the solution on the following url  http://www.xyz.com/endpoint/(????what should i write here???)
What should i add to it to be a full URL

**Reply by gabrieldwight** [reply]
something like: http://www.xyz.com/chatbot where **chatbot** is the route you have defined in your API project

**Reply by MasterBehind** [reply]
ok, now it is clear, many thanks again

**Comment by MasterBehind** [comment]
Hi,
i used conveyor extension, and run the debug, and was able to see the site, but when tried to send message i got the following error as dispalyed in the attached image
<img width="923" alt="image" src="https://github.com/gabrieldwight/Whatsapp-Business-Cloud-Api-Net/assets/152811181/ccf5681d-23ab-4988-a35f-f1d98a9508f6">

**Reply by gabrieldwight** [reply, marked as answer]
Check your whatsapp cloud api credentials and make sure the access token is not expired

**Comment by MasterBehind** [comment]
Thanks a lot, i will check. and also it will be great if i can contact you on whatsapp, my number is 0096597205246 for some business opportunity in the same domain. it will be great if you contact me.

**Comment by MasterBehind** [comment]
Thanks a lot, i will check. and also it will be great if i can contact you on whatsapp, my number is 0096597205246 for some business opportunity in the same domain. it will be great if you contact me.

**Comment by MasterBehind** [comment]
Thanks a lot, i will check. and also it will be great if i can contact you on whatsapp, my number is 0096597205246 for some business opportunity in the same domain. it will be great if you contact me.

**Comment by theCJMan** [comment]
Hi there
I am a bit new to this web dev stuff..
I tried to read all of the articles and items in the FAQ here.

Im still stuck..
If I understand it correctly, the sample app does NOT have a sample / working WebHook code section..

![image](https://github.com/gabrieldwight/Whatsapp-Business-Cloud-Api-Net/assets/38073035/5783dfda-4a03-4c4a-aece-c22f38ab420f)

If my WebHook URL is: http://www.MySite.com/Webhook
and the token (I assume I use any token??) is ASDfqweR12345
then from what I read, it seems I have to place some code in Program.cs like the below
```
app.MapControllerRoute(name: "Webhook",
                pattern: "webhook/",
                defaults: new { controller = "Webhook"});
```

then there should be a controller called WebhookController.cs

In this file I should have 
```
public class WebhookController : Controller
{
    [Route("Home/Webhook")]
    public IActionResult Webhook()
    {
       // Webhook code
    }

   [HttpGet("Home/Webhook")]
    public ActionResult<string> ConfigureWhatsAppMessageWebhook([FromQuery(Name = "hub.mode")] string hubMode,
                                                                    [FromQuery(Name = "hub.challenge")] int hubChallenge,
                                                                    [FromQuery(Name = "hub.verify_token")] string hubVerifyToken)
    {
      return Ok(hubChallenge);
     }

    [HttpPost("Home/Webhook")]
    public IActionResult ReceiveWhatsAppTextMessage([FromBody] dynamic messageReceived)
    {
    // Logic to handle different type of messages received
     return Ok();
     }
}

```

**Reply by gabrieldwight** [reply]
For minimal api since I have not done projects with it, you have to place the webhook code to program.cs file. The sample app does not have a working webhook code for testing since you have to configure it with any value for verifytoken for it to be active from the subscription options provided by Meta and the project has to be online when debugging for it to work other way you can simulate with postman and get the webhook payload from the documentation. I only had code when I was testing webhook with a local tunnel sent by Meta. 

[Webhook documentation](https://developers.facebook.com/docs/whatsapp/cloud-api/webhooks/components/#webhooks-notification-payload-reference)
[Webhook payload examples](https://developers.facebook.com/docs/whatsapp/cloud-api/webhooks/payload-examples)

**Comment by theCJMan** [comment]
I have fixed it and built some methods in the sample app.
When i make a pull request, you can check if you want to include it in the main of your repo

Thanx

⁣Get BlueMail for Android ​
